### PR TITLE
Ban direct indexing

### LINKED ISSUE #278: lints to prevent panicking + direct indexing
Need a lint (or something) preventing any kind of panicking or direct array indexing in pallets

AC:
- [ ] no longer possible to panic in a pallet without CI failure

### PR DESCRIPTION
**Pull request by keithtensor**
This PR resolves #278 by enabling CI to detect direct indexing and calls to `unwrap`.

### CODE REVIEW

**Comment by sam0x17** [comment]
note for reviewers: we are manually adding `#[allow(clippy::indexing_slicing)]` in a few places for now until we have a sparse matrix implementation that can replace these few places where this is still unavoidable. Luckily we can global find for `#[allow(clippy::indexing_slicing)]` when the time comes, just wanted to document that here

File: node/src/chain_spec.rs [hunk 1]
```diff
@@ -1,3 +1,6 @@
+// Allowed since it's actually better to panic during chain setup when there is an error
+#![allow(clippy::unwrap_used)]
```
Comment (orriin) [inline node/src/chain_spec.rs:2]: Yeah, nice to allow this in tests, benchmarks, and some node stuff like this

File: pallets/registry/src/benchmarking.rs [hunk 2]
```diff
@@ -40,7 +41,7 @@ mod benchmarks {
     fn set_identity() {
         // The target user
         let caller: T::AccountId = whitelisted_caller();
-        let _ = T::Currency::make_free_balance_be(&caller, BalanceOf::<T>::max_value());
+        let _ = T::Currency::set_balance(&caller, BalanceOf::<T>::max_value());
```
Comment (orriin) [inline pallets/registry/src/benchmarking.rs:44]: Why we have to change these things here?

File: pallets/registry/src/lib.rs [hunk 3]
```diff
@@ -125,11 +137,21 @@ pub mod pallet {
             let old_deposit = id.deposit;
             id.deposit = T::InitialDeposit::get() + fd;
             if id.deposit > old_deposit {
-                T::Currency::reserve(&who, id.deposit - old_deposit)?;
+                T::Currency::hold(
+                    &HoldReason::RegistryIdentity.into(),
+                    &who,
+                    id.deposit - old_deposit,
+                )?;
```
Comment (orriin) [inline pallets/registry/src/lib.rs:144]: Is there a reason to do the currency -> fungible migration here? Seems unrelated to direct indexing and I think you need a migration?

File: pallets/subtensor/tests/block_step.rs [hunk 4]
```diff
@@ -5,6 +5,7 @@ use mock::*;
 use sp_core::U256;
 
 #[test]
+#[allow(clippy::unwrap_used)]
```
Comment (distributedstatemachine) [inline pallets/subtensor/tests/block_step.rs:8]: Does it make sense to allow unwraps for all the tests ?

File: pallets/subtensor/src/registration.rs [hunk 5]
```diff
@@ -529,134 +529,36 @@ impl<T: Config> Pallet<T> {
         hash_as_vec
     }
 
-    pub fn hash_block_and_hotkey(block_hash_bytes: &[u8], hotkey: &T::AccountId) -> H256 {
+    #[allow(clippy::indexing_slicing)]
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/registration.rs:532]: Is this something you plan of tackling when you deal with Sparse matrix calculation i.e. https://github.com/orgs/opentensor/projects/27/views/9?pane=issue&itemId=59053066

**Review by distributedstatemachine — APPROVED** [review]

File: pallets/subtensor/src/registration.rs (hunk 5 above)
Comment (keithtensor) [reply pallets/subtensor/src/registration.rs:532]: correct.

File: pallets/subtensor/tests/block_step.rs (hunk 4 above)
Comment (keithtensor) [reply pallets/subtensor/tests/block_step.rs:8]: yes, tests don't affect the runtime, so it can use unwraps or direct indexing.

File: pallets/registry/src/lib.rs (hunk 3 above)
Comment (keithtensor) [reply pallets/registry/src/lib.rs:144]: It should have gone into the 1.0 branch, but I forgot to commit it and now i'm rolling into this.

Comment (distributedstatemachine) [reply pallets/registry/src/lib.rs:144]: @gregzaitsev  is this what you were referring to a while back ? I remember we had some conversation about it. https://github.com/opentensor/subtensor/issues/353

File: pallets/subtensor/src/math.rs [hunk 6]
```diff
@@ -270,34 +257,32 @@ pub fn inplace_normalize_64(x: &mut [I64F64]) {
...
+    x.iter()
+        .zip(y)
+        .map(|(x_i, y_i)| {
+            if *y_i != 0 {
+                x_i / y_i
+            } else {
+                I32F32::from_num(0)
+            }
+        })
+        .collect()
```
Comment (sam0x17) [inline pallets/subtensor/src/math.rs:276]: nice

**Review by sam0x17 — APPROVED** [review]
💯 awesome
